sw/device/lib/dif/dif_usbdev.c

```diff
@@ -0,0 +1,1061 @@
+ * there is a hardware change.
+ */
+_Static_assert(USBDEV_NUM_ENDPOINTS == USBDEV_PARAM_NENDPOINTS,
+               "Mismatch in number of endpoints");
+
+/**
+ * Max packet size is equal to the size of buffer entries.
+ */
+#define USBDEV_BUFFER_ENTRY_SIZE_BYTES USBDEV_MAX_PACKET_SIZE
+
+/**
+ * Bit index for the state of read buffer operation in
+ * dif_usbdev_t.active_buffer_ops`. See the definition of `buffer_op_init`.
+ */
+#define USBDEV_READ_BUFFER_OP_BIT_INDEX 15
+
+/**
+ * Constants used to indicate that a buffer pool is full or empty.
+ */
+#define BUFFER_POOL_FULL (USBDEV_NUM_BUFFERS - 1)
+#define BUFFER_POOL_EMPTY -1
+
+/**
+ * Hardware information for endpoints.
+ */
+typedef struct endpoint_hw_info {
+  uint32_t config_in_reg_offset;
+  uint8_t bit_index;
+} endpoint_hw_info_t;
+
+/**
+ * Helper macro to define an `endpoint_hw_info_t` entry for endpoint N.
+ *
+ * Note: This uses the bit indices of `USBDEV_IN_SENT` register for the sake
+ * of conciseness because other endpoint registers use the same layout.
+ */
+#define ENDPOINT_HW_INFO_ENTRY(N)                                 \
+  [N] = {.config_in_reg_offset = USBDEV_CONFIGIN##N##_REG_OFFSET, \
+         .bit_index = USBDEV_IN_SENT_SENT##N}
+
+static const endpoint_hw_info_t kEndpointHwInfos[USBDEV_NUM_ENDPOINTS] = {
+    ENDPOINT_HW_INFO_ENTRY(0),  ENDPOINT_HW_INFO_ENTRY(1),
+    ENDPOINT_HW_INFO_ENTRY(2),  ENDPOINT_HW_INFO_ENTRY(3),
+    ENDPOINT_HW_INFO_ENTRY(4),  ENDPOINT_HW_INFO_ENTRY(5),
+    ENDPOINT_HW_INFO_ENTRY(6),  ENDPOINT_HW_INFO_ENTRY(7),
+    ENDPOINT_HW_INFO_ENTRY(8),  ENDPOINT_HW_INFO_ENTRY(9),
+    ENDPOINT_HW_INFO_ENTRY(10), ENDPOINT_HW_INFO_ENTRY(11),
+};
+
+#undef ENDPOINT_HW_INFO_ENTRY
+
+/**
+ * Mapping from `dif_usbdev_irq_t` to bit indices in interrupt registers.
+ */
+static const uint8_t kIrqEnumToBitIndex[] = {
+        [kDifUsbdevIrqPktReceived] = USBDEV_INTR_COMMON_PKT_RECEIVED,
+        [kDifUsbdevIrqPktSent] = USBDEV_INTR_COMMON_PKT_SENT,
+        [kDifUsbdevIrqDisconnected] = USBDEV_INTR_COMMON_DISCONNECTED,
+        [kDifUsbdevIrqHostLost] = USBDEV_INTR_COMMON_HOST_LOST,
+        [kDifUsbdevIrqLinkReset] = USBDEV_INTR_COMMON_LINK_RESET,
+        [kDifUsbdevIrqLinkSuspend] = USBDEV_INTR_COMMON_LINK_SUSPEND,
+        [kDifUsbdevIrqLinkResume] = USBDEV_INTR_COMMON_LINK_RESUME,
+        [kDifUsbdevIrqAvEmpty] = USBDEV_INTR_COMMON_AV_EMPTY,
+        [kDifUsbdevIrqRxFull] = USBDEV_INTR_COMMON_RX_FULL,
+        [kDifUsbdevIrqAvOverflow] = USBDEV_INTR_COMMON_AV_OVERFLOW,
+        [kDifUsbdevIrqLinkInError] = USBDEV_INTR_COMMON_LINK_IN_ERR,
+        [kDifUsbdevIrqRxCrcError] = USBDEV_INTR_COMMON_RX_CRC_ERR,
+        [kDifUsbdevIrqRxPidError] = USBDEV_INTR_COMMON_RX_PID_ERR,
+        [kDifUsbdevIrqRxBitstuffError] = USBDEV_INTR_COMMON_RX_BITSTUFF_ERR,
+        [kDifUsbdevIrqFrame] = USBDEV_INTR_COMMON_FRAME,
+        [kDifUsbdevIrqConnected] = USBDEV_INTR_COMMON_CONNECTED,
+};
```

Comment:
Most other DIFs opt for a helper function with a `switch`. Is there a reason this approach is better in this case?

Reply:
I think separating validation and mapping of irqs resulted in simpler call sites in this case.

Reply:
I'm not gonna force you to change it, but I would consider switching to a function with a switch statement in it just to be consistent with the rest of the DIFs.

Is there a reason you really want to do an enum bounds check instead of catching it in a switch?

Reply:
I can see how in this case, doing it with an array makes it simpler.

Let's leave it like this and come back to it. USB has way more interrupts than quite a few other DIFs, and I'm not sure we fully know if we're going to move to a mask-based public interface for interrupts anyway.

Reply:
We could also consider changing the IP to lump all the error conditions into a single interrupt line with a separate error status?

Since the hardware does not match anything existing, we could certainly use driver writer's experience to change things.

Reply:
@mdhayter If I understand correctly, it would simplify the hw interface of usbdev but wouldn't really affect this part since callers would still be interested in individual bits.